### PR TITLE
fby3: dl: support 1ou vpp

### PR DESCRIPTION
**Pull request by Eddie043**
Summary:

- Support VF VPP function
- Support command to get 1ou expansion board type

Test Plan:

Build code: PASS
Check 1ou type: PASS
Check sensor read: PASS
Check VPP log: PASS

Log
- Get 1ou board type

root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x00 0x00
9C 9C 00 04

- Toggle VPP and check sensor read/VPP log

root@bmc-oob:~# bic-util slot1 0x18 0x52 0x05 0x42 0x00 0x0F 0xE1
root@bmc-oob:~# bic-util slot1 0x18 0x52 0x05 0x42 0x00 0x0F 0xFF
root@bmc-oob:~# sensor-util slot1 | grep -i e1s
E1S Outlet Temp              (0x50) :   30.00 C     | (ok)
E1S P12V_AUX                 (0x51) :   12.21 Volts | (ok)
E1S P12V_EDGE                (0x52) :   12.15 Volts | (ok)
E1S P3V3_AUX                 (0x53) :    3.28 Volts | (ok)
E1S DEV0 Power               (0x60) : NA | (na)
E1S DEV0 Voltage             (0x61) : NA | (na)
E1S DEV0 Temp                (0x62) : NA | (na)
E1S 3V3_ADC_DEV0             (0x64) : NA | (na)
E1S 12V_ADC_DEV0             (0x63) : NA | (na)
E1S DEV1 Power               (0x68) : NA | (na)
E1S DEV1 Voltage             (0x69) : NA | (na)
E1S DEV1 Temp                (0x6A) : NA | (na)
E1S 3V3_ADC_DEV1             (0x6C) : NA | (na)
E1S 12V_ADC_DEV1             (0x6B) : NA | (na)
E1S DEV2 Power               (0x70) : NA | (na)
E1S DEV2 Voltage             (0x71) : NA | (na)
E1S DEV2 Temp                (0x72) : NA | (na)
E1S 3V3_ADC_DEV2             (0x74) : NA | (na)
E1S 12V_ADC_DEV2             (0x73) : NA | (na)
E1S DEV3 Power               (0x78) : NA | (na)
E1S DEV3 Voltage             (0x79) : NA | (na)
E1S DEV3 Temp                (0x7A) : NA | (na)
E1S 3V3_ADC_DEV3             (0x7C) : NA | (na)
E1S 12V_ADC_DEV3             (0x7B) : NA | (na)

root@bmc-oob:~# bic-util slot1 0x18 0x52 0x05 0x42 0x00 0x0F 0xE1
root@bmc-oob:~# bic-util slot1 0x18 0x52 0x05 0x42 0x00 0x0F 0xFF
root@bmc-oob:~# sensor-util slot1 | grep -i e1s
E1S Outlet Temp              (0x50) :   30.00 C     | (ok)
E1S P12V_AUX                 (0x51) :   12.21 Volts | (ok)
E1S P12V_EDGE                (0x52) :   12.15 Volts | (ok)
E1S P3V3_AUX                 (0x53) :    3.28 Volts | (ok)
E1S DEV0 Power               (0x60) :    2.80 Watts | (ok)
E1S DEV0 Voltage             (0x61) :   12.15 Volts | (ok)
E1S DEV0 Temp                (0x62) :   31.00 C     | (ok)
E1S 3V3_ADC_DEV0             (0x64) :    3.30 Volts | (ok)
E1S 12V_ADC_DEV0             (0x63) :   12.09 Volts | (ok)
E1S DEV1 Power               (0x68) :    2.80 Watts | (ok)
E1S DEV1 Voltage             (0x69) :   12.15 Volts | (ok)
E1S DEV1 Temp                (0x6A) :   31.00 C     | (ok)
E1S 3V3_ADC_DEV1             (0x6C) :    3.30 Volts | (ok)
E1S 12V_ADC_DEV1             (0x6B) :   12.03 Volts | (ok)
E1S DEV2 Power               (0x70) :    2.80 Watts | (ok)
E1S DEV2 Voltage             (0x71) :   12.15 Volts | (ok)
E1S DEV2 Temp                (0x72) :   31.00 C     | (ok)
E1S 3V3_ADC_DEV2             (0x74) :    3.28 Volts | (ok)
E1S 12V_ADC_DEV2             (0x73) :   12.09 Volts | (ok)
E1S DEV3 Power               (0x78) :    2.70 Watts | (ok)
E1S DEV3 Voltage             (0x79) :   12.15 Volts | (ok)
E1S DEV3 Temp                (0x7A) :   30.00 C     | (ok)
E1S 3V3_ADC_DEV3             (0x7C) :    3.32 Volts | (ok)
E1S 12V_ADC_DEV3             (0x7B) :   12.03 Volts | (ok)

root@bmc-oob:~# log-util slot1 --print
2018 Jan 01 00:25:31 log-util: User cleared FRU: 1 logs
TATUS (0x46), Event Data: (0B013A) 1OU/Num 0 VPP Power Control Triggered
1    slot1    2018-01-01 01:06:31    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:06:31, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013B) 1OU/Num 1 VPP Power Control Triggered
1    slot1    2018-01-01 01:06:35    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:06:35, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013C) 1OU/Num 2 VPP Power Control Triggered
1    slot1    2018-01-01 01:06:36    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:06:36, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013D) 1OU/Num 3 VPP Power Control Triggered
1    slot1    2018-01-01 01:08:08    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:08:08, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013C) 1OU/Num 2 VPP Power Control Triggered
1    slot1    2018-01-01 01:08:08    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:08:08, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013B) 1OU/Num 1 VPP Power Control Triggered
1    slot1    2018-01-01 01:08:08    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:08:08, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013D) 1OU/Num 3 VPP Power Control Triggered
1    slot1    2018-01-01 01:08:08    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2018-01-01 01:08:08, Sensor: SYSTEM_STATUS (0x46), Event Data: (0B013A) 1OU/Num 0 VPP Power Control Triggered

### CODE REVIEW

File: meta-facebook/yv3-dl/src/platform/plat_isr.c [hunk 1]
```diff
@@ -583,3 +594,103 @@ void ISR_SMI()
...
+	I2C_MSG i2c_msg = { 0 };
+	i2c_msg.bus = I2C_BUS2;
+	i2c_msg.target_addr = CPLD_ADDR;
+	i2c_msg.data[0] = CPLD_1OU_VPP_POWER_STATUS;
+	i2c_msg.tx_len = 1;
+	i2c_msg.rx_len = 1;
+
+	const uint8_t retry = 3;
+	if (i2c_master_read(&i2c_msg, retry)) {
+		LOG_ERR("Failed to read CPU VPP status, bus0x%x addr0x%x offset0x%x\n", i2c_msg.bus,
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_isr.c:657]: Newline (`\n`) not necessary

File: meta-facebook/yv3-dl/src/platform/plat_isr.c [hunk 2]
```diff
@@ -583,3 +594,103 @@ void ISR_SMI()
...
+		common_addsel_msg_t sel_msg = { 0 };
+		sel_msg.InF_target = BMC_IPMB;
+		sel_msg.sensor_type = IPMI_OEM_SENSOR_TYPE_SYS_STA;
+		sel_msg.event_type = IPMI_OEM_EVENT_TYPE_NOTIFY;
+		sel_msg.sensor_number = SENSOR_NUM_SYS_STA;
+		sel_msg.event_data1 = IPMI_OEM_EVENT_OFFSET_VPP_EVENT;
+		sel_msg.event_data2 = IPMI_OEM_EVENT_OFFSET_1OU;
+		sel_msg.event_data3 = _1ou_m2_name_mapping_table[device_id];
+		if (!common_add_sel_evt_record(&sel_msg))
+			LOG_ERR("addsel fail\n");
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_isr.c:692]: LOG_XXX adds a newline automatically, no need to add one.

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 3]
```diff
@@ -105,6 +202,82 @@ void init_platform_config()
...
+				if (_1ou_status.card_type != TYPE_1OU_UNKNOWN) {
+					tx_len = 2;
+					rx_len = 0;
+					memset(data, 0, I2C_DATA_SIZE);
+					data[0] = CPLD_1OU_CARD_DETECTION;
+					data[1] = _1ou_status.card_type;
+					i2c_msg = construct_i2c_message(I2C_BUS1, CPLD_ADDR, tx_len,
+									data, rx_len);
+					if (i2c_master_write(&i2c_msg, retry)) {
+						printf("Failed to set 1OU card detection to CPLD register(0x%x)\n",
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:267]: remove newline character.
Change to LOG_ERR(...)

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 4]
```diff
@@ -26,15 +27,110 @@
...
+	reg_value = sys_read32(AST1030_ADC_BASE_ADDR + engine_control);
+	switch (reg_value & (BIT(7) | BIT(6))) {
+	case REF_VOL_2_5V:
+		reference_voltage = 2.5;
+		break;
+	case REF_VOL_1_2V:
+		reference_voltage = 1.2;
+		break;
+	default:
+		printf("Unsupported the external reference voltage\n");
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:120]: remove newline character

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 5]
```diff
@@ -26,15 +27,110 @@
...
+	{ 1.5, RANGE, TYPE_1OU_EXP_WITH_NIC },
+	{ 1.75, RANGE, TYPE_1OU_VERNAL_FALLS_WITH_AST },
+};
+
+bool get_adc_voltage(int channel, float *voltage)
+{
+	CHECK_NULL_ARG_WITH_RETURN(voltage, false);
+
+	if (channel >= NUMBER_OF_ADC_CHANNEL) {
+		printf("Invalid ADC channel-%d\n", channel);
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:98]: Change to LOG_ERR(...)

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 6]
```diff
@@ -87,6 +183,7 @@ void init_platform_config()
 	i2c_msg = construct_i2c_message(I2C_BUS2, CPLD_ADDR, tx_len, data, rx_len);
 	if (!i2c_master_read(&i2c_msg, retry)) {
 		class_type = i2c_msg.data[0];
+		printf("class_type = %x\n", class_type);
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:186]: Should be LOG_INF(...)

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 7]
```diff
@@ -105,6 +202,82 @@ void init_platform_config()
...
+					if ((voltage >
+					     typical_voltage - (typical_voltage * 0.05)) &&
+					    (voltage <
+					     typical_voltage + (typical_voltage * 0.05))) {
+						_1ou_status.card_type =
+							_1ou_card_mapping_table[cnt].card_type;
+					}
+					break;
+				default:
+					printf("Unknown condition 0x%x",
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:253]: LOG_ERR(...)

File: meta-facebook/yv3-dl/src/platform/plat_class.c [hunk 8]
```diff
@@ -105,6 +202,82 @@ void init_platform_config()
...
+									data, rx_len);
+					if (i2c_master_write(&i2c_msg, retry)) {
+						printf("Failed to set 1OU card detection to CPLD register(0x%x)\n",
+						       data[0]);
+					}
+					break;
+				}
+				if ((cnt == ARRAY_SIZE(_1ou_card_mapping_table)) &&
+				    (_1ou_status.card_type == TYPE_1OU_UNKNOWN)) {
+					printf("Unknown the 1OU card type, the voltage of ADC channel-6 is %fV\n",
```
Comment (GoldenBug) [inline meta-facebook/yv3-dl/src/platform/plat_class.c:274]: LOG_ERR(...).
Remove newline

**Comment by Eddie043** [comment]
Hi @GoldenBug,
Thanks for the suggestion, I followed it for a new commit, please take a look.

**Pull request closed without merging**